src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/SnapshotTracker.kt

```diff
@@ -0,0 +1,231 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2024 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.jetbrains.kotlinx.lincheck.strategy.managed
+
+import org.jetbrains.kotlinx.lincheck.findField
+import org.jetbrains.kotlinx.lincheck.getFieldOffset
+import org.jetbrains.kotlinx.lincheck.isPrimitiveWrapper
+import org.jetbrains.kotlinx.lincheck.strategy.managed.SnapshotTracker.Descriptor.*
+import org.jetbrains.kotlinx.lincheck.strategy.managed.SnapshotTracker.MemoryNode.*
+import org.jetbrains.kotlinx.lincheck.util.*
+import java.lang.reflect.Field
+import java.lang.reflect.Modifier
+import java.util.Collections
+import java.util.IdentityHashMap
+import java.util.concurrent.atomic.AtomicIntegerArray
+import java.util.concurrent.atomic.AtomicLongArray
+import java.util.concurrent.atomic.AtomicReferenceArray
+
+
+/**
+ * Manages a snapshot of the global static state.
+ *
+ * This class only tracks static memory and memory reachable from it,
+ * referenced from the test code. So the whole static state is not recorded, but only a subset of that,
+ * which is named *snapshot*.
+ */
+class SnapshotTracker {
+    private val trackedObjects = IdentityHashMap<Any, MutableList<MemoryNode>>()
+
+    private sealed class Descriptor {
+        class FieldDescriptor(val field: Field, val offset: Long) : Descriptor()
+        class ArrayCellDescriptor(val index: Int) : Descriptor()
+    }
+
+    private sealed class MemoryNode(
+        val descriptor: Descriptor,
+        val initialValue: Any?
+    ) {
+        class RegularFieldNode(descriptor: FieldDescriptor, initialValue: Any?) : MemoryNode(descriptor, initialValue)
+        class StaticFieldNode(descriptor: FieldDescriptor, initialValue: Any?) : MemoryNode(descriptor, initialValue)
+        class ArrayCellNode(descriptor: ArrayCellDescriptor, initialValue: Any?) : MemoryNode(descriptor, initialValue)
+    }
+
+    fun trackField(obj: Any?, className: String, fieldName: String) {
+        if (obj != null && obj !in trackedObjects) return
+
+        val clazz: Class<*> = Class.forName(className).let {
+            if (obj != null) it
+            else it.findField(fieldName).declaringClass
+        }
+        val field = clazz.findField(fieldName)
+        val readResult = runCatching { readFieldViaUnsafe(obj, field) }
+
+        if (readResult.isSuccess) {
+            val fieldValue = readResult.getOrNull()
+            trackSingleField(obj, clazz, field, fieldValue) {
+                if (shouldTrackEnergetically(fieldValue)) {
```

Comment:
I think, a more widely used (in programming) antonym of `lazy` is `eager`, not `energetic`.